src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp

```diff
@@ -378,88 +421,101 @@ FixedwingPositionControl::calculate_target_airspeed(float airspeed_demand, const
+
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	case TECS::ECL_TECS_MODE_BAD_DESCENT:
-		t.mode = tecs_status_s::TECS_MODE_BAD_DESCENT;
 		break;
 
-	case TECS::ECL_TECS_MODE_CLIMBOUT:
-		t.mode = tecs_status_s::TECS_MODE_CLIMBOUT;
+	case ECO_FULL:
+		if (_conditions_for_eco_dash_met) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_ECO;
+
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
+
 		break;
-	}
 
-	t.altitude_sp = _tecs.hgt_setpoint();
-	t.altitude_filtered = _tecs.vert_pos_state();
+	case DASH_CRUISE:
+		if (_conditions_for_eco_dash_met && _tecs.get_flight_phase() == tecs_status_s::TECS_FLIGHT_PHASE_LEVEL) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_DASH;
 
-	t.true_airspeed_sp = _tecs.TAS_setpoint_adj();
-	t.true_airspeed_filtered = _tecs.tas_state();
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	t.height_rate_setpoint = _tecs.hgt_rate_setpoint();
-	t.height_rate = _tecs.vert_vel_state();
+		break;
 
-	t.equivalent_airspeed_sp = _tecs.get_EAS_setpoint();
-	t.true_airspeed_derivative_sp = _tecs.TAS_rate_setpoint();
-	t.true_airspeed_derivative = _tecs.speed_derivative();
-	t.true_airspeed_derivative_raw = _tecs.speed_derivative_raw();
-	t.true_airspeed_innovation = _tecs.getTASInnovation();
+	case DASH_FULL:
+		if (_conditions_for_eco_dash_met) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_DASH;
 
-	t.total_energy_error = _tecs.STE_error();
-	t.total_energy_rate_error = _tecs.STE_rate_error();
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	t.energy_distribution_error = _tecs.SEB_error();
-	t.energy_distribution_rate_error = _tecs.SEB_rate_error();
+		break;
+	}
+}
 
-	t.total_energy = _tecs.STE();
-	t.total_energy_rate = _tecs.STE_rate();
-	t.total_energy_balance = _tecs.SEB();
-	t.total_energy_balance_rate = _tecs.SEB_rate();
+void
+FixedwingPositionControl::check_eco_dash_allowed()
+{
+	const float altitude_amsl_min = max(_tecs.get_hgt_setpoint() - fw_alt_err_u.get(),
+					    _local_pos.ref_alt + fw_alt_min.get());
+	const float altitude_amsl_max = _tecs.get_hgt_setpoint() + fw_alt_err_o.get();
 
-	t.total_energy_sp = _tecs.STE_setpoint();
-	t.total_energy_rate_sp = _tecs.STE_rate_setpoint();
-	t.total_energy_balance_sp = _tecs.SEB_setpoint();
-	t.total_energy_balance_rate_sp = _tecs.SEB_rate_setpoint();
+	const bool altitdue_conditions_met = _current_altitude <= altitude_amsl_max
+					     && _current_altitude >= altitude_amsl_min;
 
-	t.throttle_integ = _tecs.throttle_integ_state();
-	t.pitch_integ = _tecs.pitch_integ_state();
+	// add a 10s timeout after conditions where not met
+	_conditions_for_eco_dash_met = altitdue_conditions_met && hrt_elapsed_time(&_time_conditions_not_met) > 10_s;
 
-	t.throttle_sp = _tecs.get_throttle_setpoint();
-	t.pitch_sp_rad = _tecs.get_pitch_setpoint();
 
-	t.timestamp = hrt_absolute_time();
+	if (!altitdue_conditions_met) {
```

Comment:
@sfuhrer If you put this code above the line where you evaluate `_conditions_for_eco_dash_met` then you can just write
`_conditions_for_eco_dash_met = hrt_elapsed_time(&_time_conditions_not_met) > 10_s`

Reply:
:+1